com.unity.netcode.gameobjects/Runtime/Utility/DynamicUnmanagedArray.cs

```diff
@@ -0,0 +1,104 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+
+namespace Unity.Netcode
+{
+    public struct DynamicUnmanagedArray<T> : IReadOnlyList<T>, IDisposable where T : unmanaged
```

Comment:
In addition to where it's currently used, I'm also intending to update FastBufferReader/FastBufferWriter to be able to use this. The big benefit is that length and capacity are dissociated here and the length value can be changed, so this would give us the ability to have messages with dynamic-length data that could be written directly into this array. This would mean we could reuse one array for reading a given piece of data and it'd be able to grow if needed (and FastBufferReader could do that automatically) and if it doesn't need to grow, FastBufferReader could read a variable amount of data and update the length value without having any allocations... and when it grows it would allocate through Malloc instead of GC...

Downside, though, is that it has to be manually freed via Dispose() when it's no longer being used. So I'm open for feedback on whether something like this is useful/good, or too dangerous to be willing to accept.